multi_partner_learning.py

```diff
@@ -73,54 +82,151 @@ def __init__(self,
 
         logger.debug("MultiPartnerLearning object instantiated.")
 
+    def train_single_model(self, partner):
+        # first check if the model was not already trained
+        if partner.model == None: 
+            # Initialize model
+            model = utils.generate_new_cnn_model()
+
+            # Set if early stopping if needed
+            cb = []
+            if self.is_early_stopping:
+                es = EarlyStopping(
+                    monitor="val_loss",
+                    mode="min",
+                    verbose=0,
+                    patience=constants.PATIENCE,
+                )
+                cb.append(es)
+
+            # Train model
+            logger.info(f"   Training model on partner with id #{partner.id}")
+            model.fit(
+                partner.x_train,
+                partner.y_train,
+                batch_size=partner.batch_size,
+                epochs=self.epoch_count,
+                verbose=0,
+                validation_data=self.val_data,
+                callbacks=cb,
+            )
+            # store the model
+            partner.model = model
+
+            self.nb_epochs_done = (
+                (es.stopped_epoch + 1) if es.stopped_epoch != 0 else self.epoch_count
+            )
+
+            return model
+        else:
+            return partner.model
+
+    def make_stacked_meta_model(self, meta_model_hidden_dim=10):
+        logger.info("## Making metamodel")
+        # make a list with single partner 's models
+        list_of_models = [None] * self.partners_count
+        for partner_idx, partner in enumerate(self.partners_list):
+            list_of_models[partner_idx] = self.train_single_model(partner)
+
+        # update all layers in all models to not be trainable
+        for i in range(len(list_of_models)):
+            model = list_of_models[i]
+            for layer in model.layers:
+                # make not trainable
+                layer.trainable = False
+                # rename to avoid 'unique layer name' issue
+                layer.name = "ensemble_" + str(i + 1) + "_" + layer.name
+        # define multi-headed input
+        ensemble_visible = [model.input for model in list_of_models]
+        # concatenate merge output from each model
+        ensemble_outputs = [model.output for model in list_of_models]
+        merge = concatenate(ensemble_outputs)
+        hidden = Dense(meta_model_hidden_dim, activation="relu")(merge)
+        output = Dense(constants.NUM_CLASSES, activation="softmax")(hidden)
+        meta_model = Model(inputs=ensemble_visible, outputs=output)
+        # compile
+        meta_model.compile(
+            loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"]
+        )
+        return meta_model
```

Comment:
```suggestion
    def make_stacked_meta_model(self, meta_model_hidden_dim=10):
    
        logger.info("## Model ensembling approach - Creating the meta-model")
        
        # Create a list with an independently trained model for each partner
        list_of_models = []
        for partner in self.partners_list:
            list_of_models.append(self.train_single_model(partner))

        # Update all layers in all models to not be trainable
        for i in range(len(list_of_models)):
            model = list_of_models[i]
            for layer in model.layers:
                layer.trainable = False  # Make it not trainable
                layer.name = f"ensemble_{i+1}_{layer.name}"  # Rename it to avoid 'unique layer name' issue
                
        # Define multi-headed input
        ensemble_visible = [model.input for model in list_of_models]
        
        # Concatenate merge output from each model
        ensemble_outputs = [model.output for model in list_of_models]
        merge = concatenate(ensemble_outputs)
        hidden = Dense(meta_model_hidden_dim, activation="relu")(merge)
        output = Dense(constants.NUM_CLASSES, activation="softmax")(hidden)
        meta_model = Model(inputs=ensemble_visible, outputs=output)
        
        # Compile the meta-model
        meta_model.compile(
            loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"]
        )
        
        return meta_model
```